pkg/operator/certrotation/signer_test.go

```diff
@@ -1,24 +1,219 @@
+			// wait until the first controller finishes
+			// so that we can update the secret with the new keys
+			<-firstControllerEndChan
+
+			// cannot use fakeKubeClient.ActionsA()
+			// because it deadlocks, fakeKubeClient.RLock already acquired
+			//
+			// the created secret from the actions
+			// is validated at the end of this test
+			// and must match this one
+			createdSecret := existingSecret.DeepCopy()
+			createdSecret.Type = corev1.SecretTypeTLS
+			createdSecret.Annotations["openshift.io/owning-component"] = "test"
+			return true, createdSecret, nil
+		}
+	}())
+
+	c := &RotatedSigningCASecret{
+		Namespace:     "ns",
+		Name:          "signer",
+		Validity:      24 * time.Hour,
+		Refresh:       12 * time.Hour,
+		Client:        fakeKubeClient.CoreV1(),
+		Lister:        corev1listers.NewSecretLister(indexer),
+		EventRecorder: events.NewInMemoryRecorder("test"),
+		AdditionalAnnotations: AdditionalAnnotations{
+			JiraComponent: "test",
+		},
+		Owner: &metav1.OwnerReference{
+			Name: "operator",
+		},
+	}
+
+	var signingKeyPairFromControllerTwo *crypto.CA
+	go func() {
+		// start the second controller
+		var err error
+		<-secondControllerStartChan
+		signingKeyPairFromControllerTwo, err = c.ensureSigningCertKeyPair(context.TODO())
+		if err != nil {
+			t.Error(err)
+		}
+		close(secondControllerEndChan)
+	}()
+
+	// start the first controller
+	signingKeyPairFromControllerOne, err := c.ensureSigningCertKeyPair(context.TODO())
+	if err != nil {
+		t.Fatal(err)
+	}
+	certBytesFromControllerOne, keyBytesFromControllerOne, err := signingKeyPairFromControllerOne.Config.GetPEMBytes()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !bytes.Equal(existingSecret.Data["tls.crt"], certBytesFromControllerOne) {
+		t.Fatal("returned certificate data from controller one doesn't match existing secreted data")
+	}
+	if !bytes.Equal(existingSecret.Data["tls.key"], keyBytesFromControllerOne) {
+		t.Fatal("returned certificate key from controller one doesn't match existing secreted data")
+	}
+	close(firstControllerEndChan)
+	<-secondControllerEndChan
+
+	certBytesFromControllerTwo, keyBytesFromControllerTwo, err := signingKeyPairFromControllerTwo.Config.GetPEMBytes()
+	if err != nil {
+		t.Fatal(err)
+	}
+	// note that GetTLSCertificateConfigFromBytes (executed by the target method) check if
+	// the crypto material is not empty
+	//
+	// the following 2 checks show that the second controller
+	// created a new signing key pair!
+	if bytes.Equal(certBytesFromControllerOne, certBytesFromControllerTwo) {
+		t.Fatal("cert data from both controllers are equal")
+	}
+	if bytes.Equal(keyBytesFromControllerOne, keyBytesFromControllerTwo) {
+		t.Fatal("cert key from both controllers are equal")
+	}
+
+	actions := fakeKubeClient.Actions()
+	updateActionValidated := false
+	createActionValidated := false
+	for _, action := range actions {
+		switch action.GetVerb() {
+		case "create":
+			if createAction, ok := action.(clienttesting.CreateAction); ok {
```

Comment:
For future-proofing against silently skipping these assertions, how about failing if !ok?